waku/v2/utils/credentials.nim

```diff
@@ -0,0 +1,397 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import 
+  chronicles, options, json, strutils,
+  stew/byteutils,
+  std/[algorithm, os, sequtils, sets],
+  ./keyfile
+  
+type
+  IdentityTrapdoor* = seq[byte] #array[32, byte]
+  IdentityNullifier* = seq[byte] #array[32, byte]
+  # identity key as defined in https://hackmd.io/tMTLMYmTR5eynw2lwK9n1w?view#Membership
+  IdentitySecretHash* = seq[byte] #array[32, byte]
+  # hash of identity key as defined ed in https://hackmd.io/tMTLMYmTR5eynw2lwK9n1w?view#Membership
+  IDCommitment* = seq[byte] #array[32, byte]
+
+type IdentityCredential* = object
+  idTrapdoor*: IdentityTrapdoor
+  idNullifier*: IdentityNullifier
+  ## user's identity key (a secret key) which is selected randomly
+  ## see details in https://hackmd.io/tMTLMYmTR5eynw2lwK9n1w?view#Membership
+  idSecretHash*: IdentitySecretHash
+  # hash of user's identity key generated by
+  # Poseidon hash function implemented in rln lib
+  # more details in https://hackmd.io/tMTLMYmTR5eynw2lwK9n1w?view#Membership
+  idCommitment*: IDCommitment
+
+type MembershipIndex* = uint
+
+type MembershipContract* = object
+  chainId*: string
+  address*: string
```

Comment:
```suggestion
type ContractReference* = object
  chainId*: string
  address*: string
```

Reply:
No strong opinion here, but MembershipContract can refer to the type of the contract itself (generated by nim-web3). This works too!